### ISSUE TITLE
Differentiate `no-else-return` for `if` and `try` expressions

### ISSUE THREAD

**Issue by gar1t** [PyCQA/pylint#8155]
### Current problem

I can appreciate the recent checks for returns from `try` blocks:

```
try:
    x = read()
except Exception:
  return 0
else:
  return x + 1
```

However, the check is associated with `no-else-return`, as if it were equivalent to:

```
x = read()
if x is None:
  return 0
else:
  return x + 1
```

While the language designers opted for the term `else` to define the execution branch for non-exceptions, this should not imply that it is the same meaning as `else` in an `if` statement. The meaning here is quite different. Unlike `else` in an `if` statement, the `else` block for `try` is executed when the `try` block is executed, rather than when it is not.

I would argue that it's an anti-pattern to access variables defined in a `try` block from outside the block. This rule not only encourages this pattern, it requires it. I think it should have its own rule code to be disabled by those who would opt out of this.

### Desired solution

Create a new rule code for this, e.g. `no-try-else-return`, to differentiate it from the `if` statement version.

**Comment by Segelzwerg** [comment]
I agree and want to add if you raise an error in else the behavior is actually require.

```
try:
    do_something()
except SomeError:
    …
else:
    raise OtherError()
```

**Comment by RemiCardona** [comment]
Big +1 from me: splitting seems a good plan.

I have literally hundreds of such cases which I feel are justified and I would still like to retain the original `no-else-return` check as our team feels this to be very useful.

**Comment by LefterisJP** [comment]
A +1 from me. We just ugpraded to `pylint==2.16.1` and I have to now disable the `no-else-return` rule due to this. Having an else in exceptions is fine and considered good practise even.

**Comment by volans-** [comment]
FYI the same applies also for `no-else-raise`.
I'd actually wouldn't consider this an enhancement to request, but a bug to fix.
In the context of a `try-except-else` construct I don't think the `else` part should be considered bad practice if the `except` part does raise or return. It's a perfectly valid construct, and following pylint advice on this leads to brittle code that might lead to bug later on when refactoring it.

**Comment by volans-** [comment]
> > I'd actually wouldn't consider this an enhancement to request, but a bug to fix.
> 
> The expected behavior of the check is to flag any instance of `else` that is not logically required. Are there any examples of flagged `else` blocks that are logically required? That would be a bug. Otherwise, this is a request to change the expected behavior.

@nickdrozd I understand the definition, but which one of the following 2 code examples do you find more correct/pythonic/future-proof?

```python
@contextmanager
def repro(cleanup_on_error: bool = False):
    prepare()
    try:
        yield
    except SomeException:
        if cleanup_on_error:
            cleanup()
        raise
    else:
        cleanup()

    do_stuff()
```

```python
@contextmanager
def repro(cleanup_on_error: bool = False):
    prepare()
    try:
        yield
    except SomeException:
        if cleanup_on_error:
            cleanup()
        raise

    cleanup()
    do_stuff()
```

Although not logically required, IMHO the first one with the `else` is more clear and explicit. It makes clear that `cleanup()` is meant to happen when there was no exception raised  in the `try` block. If in a future refactor the `raise` will be removed, the code will naturally still be correct. While in the second code example I will most likely feel the need to add a comment to explain the relationship between `cleanup()` and the `try` block for future reference.

Hence I consider this more of a bug in the sense that requires the developer to write what I consider less clear code or forces them to add an ignore comment. But no, I don't think I've encountered any example where pylint has flagged an `else` that couldn't be logically removed.

**Comment by volans-** [comment]
@gar1t don't worry about that for me, it's clearly an enhancement ;)

**Comment by nickdrozd** [comment]
@volans- I actually do prefer the second example, because it makes it visually clear that there is just one happy path. It can be future-proofed by adding a comment that says `# DO NOT TOUCH THIS UNDER ANY CIRCUMSTANCE`. In fact, Pylint should raise a warning for any function with a `cleanup_on_error` flag that is *not* accompanied by such a comment!

But I do see your point. Probably there is no way to express it that will look nice to everyone, because error handling logic is inherently complicated and therefore unavoidably ugly. After all, it's one of the few occasions where [`goto` is considered acceptable](https://blog.joren.ga/gotophobia-harmful#errorexception-handling--cleanup).

**Comment by jacobtylerwalls** [comment]
Another voice in favor in #9008
